### PR TITLE
[TEP0138] Add Per-feature Flag Struct for New Features

### PR DESCRIPTION
**Pull request by JeromeJu**
# Changes

This commit adds the per-feature flag struct and documentations for
new features in TektonCD pipeline.

part of: TEP-0138
   

# Submitter Checklist

As the author of this PR, please check off the items in this checklist:

- [x] Has [Docs](https://github.com/tektoncd/community/blob/main/standards.md#docs) if any changes are user facing, including updates to minimum requirements e.g. Kubernetes version bumps
- [ ] Has [Tests](https://github.com/tektoncd/community/blob/main/standards.md#tests) included if any functionality added or changed
- [x] Follows the [commit message standard](https://github.com/tektoncd/community/blob/main/standards.md#commits)
- [x] Meets the [Tekton contributor standards](https://github.com/tektoncd/community/blob/main/standards.md) (including functionality, content, code)
- [x] Has a kind label. You can add one by adding a comment on this PR that contains `/kind <type>`. Valid types are bug, cleanup, design, documentation, feature, flake, misc, question, tep
- [x] Release notes block below has been updated with any user facing changes (API changes, bug fixes, changes requiring upgrade notices or deprecation warnings). See some examples of [good release notes](https://github.com/tektoncd/community/blob/main/standards.md#good-release-notes).
- [x] Release notes contains the string "action required" if the change requires additional action from users switching to the new release

# Release Notes

```release-note
Please use per feature flags for new API-driven feature gating.
```

### CODE REVIEW

**Comment by chitrangpatel** [comment]
There seems to be only one commit but the PR description says two. May be it needs to be updated.

**Comment by JeromeJu** [comment]
> There seems to be only one commit but the PR description says two. May be it needs to be updated.

Thanks @chitrangpatel . Updated!

Also to be on the note for this change:  I previously thought that we could just provide an example since there were no upcoming new features yet. But now there are a couple that I've realized: i.e. https://github.com/tektoncd/pipeline/pull/7162 and CEL expression that had been mentioned in TEP process.

Happy to pair on/ walk through the process if a new feature is getting in pipeline!

**Comment by chitrangpatel** [comment]
Do you think we can have an example of how we introduce a feature "enable-foo" in developer docs? In there we can at least provide an example of that we need to do to the struct and how it can be used internally in our codebase. Finally, when we add a new feature, we might want to also enable it in https://github.com/tektoncd/pipeline/blob/main/test/e2e-tests-kind-prow-alpha.env

I think a document showing this journey will be useful when we start adding new features under per-feature flags.

**Comment by JeromeJu** [comment]
> Do you think we can have an example of how we introduce a feature "enable-foo" in developer docs? In there we can at least provide an example of that we need to do to the struct and how it can be used internally in our codebase. Finally, when we add a new feature, we might want to also enable it in https://github.com/tektoncd/pipeline/blob/main/test/e2e-tests-kind-prow-alpha.env
> 
> I think a document showing this journey will be useful when we start adding new features under per-feature flags.

Thanks @chitrangpatel , it makes sense to add it as an example. Working on it as a 2nd commit rn.

**Comment by JeromeJu** [comment]
🙏 It'd be much appreciated if we could merge https://github.com/tektoncd/pipeline/pull/7220 first 🙏

**Review by Yongxuanzhang — APPROVED** [review]

**Comment by JeromeJu** [comment]
Since https://github.com/tektoncd/pipeline/pull/6941 has been merged and this is unblocked, while we have a handful of new features(i.e. #7277, #7255) that might come in. Could we prioritize merging this PR as the more new feature flags come into the repo the more future "patching" of the usage with the new struct there are going to be? 🙏 🙏 🙏 
cc @tektoncd/core-maintainers

File: docs/developers/feature-versioning.md [hunk 1]
```diff
@@ -0,0 +1,166 @@
+# Feature Versioning
+
+The stability levels of features (feature versioning) are independent of CRD API versions (API versioning).
+
+## Adding feature gated API fields for API-driven features title
```
Comment (QuanZhang-William) [inline docs/developers/feature-versioning.md:5]: Is this the right heading?

File: docs/developers/feature-versioning.md [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+- Write unit tests to verify the new feature flag and update all test cases that require the configMap setting, such as those related to provenance propagation.
+- To add integration tests:
+    - First, add the tests to `pull-tekton-pipeline-alpha-integration-test` by enabling the newly-introduced per-feature flag at [alpha test Prow environment](./../../test/e2e-tests-kind-prow-alpha.env).
+    - When the flag is promoted to `beta` stability level, change the test to use [beta Prow environment setup](./../../test/e2e-tests-kind-prow-beta.env).
+    - To add additional CI tests for combinations of feature flags, add tests for all alpha feature flags being turned on, with one alpha feature turned off at a time.
+- Add the tested new per-feature flag key value to the [the pipeline configMap](./../../config/config-feature-flags.yaml).
+- Update documentations for the new alpha feature at [alpha-stability-level](./../additional-configs.md#alpha-features).
+
+#### Example of adding a new Per-feature flag
```
Comment (QuanZhang-William) [inline docs/developers/feature-versioning.md:21]: It would be helpful to put a PR link as example here in the future.

File: pkg/apis/config/feature_flags.go [hunk 3]
```diff
@@ -358,3 +364,16 @@ func GetVerificationNoMatchPolicy(ctx context.Context) string {
 func IsSpireEnabled(ctx context.Context) bool {
 	return FromContextOrDefaults(ctx).FeatureFlags.EnforceNonfalsifiability == EnforceNonfalsifiabilityWithSpire
 }
+
+type PerFeatureFlag struct {
```
Comment (QuanZhang-William) [inline pkg/apis/config/feature_flags.go:368]: Can you add a TODO here as the struct is currently not used

**Comment by QuanZhang-William** [comment]
Thanks for the PR, generally looks good to me. Added some small comments.

File: pkg/apis/config/feature_flags.go (hunk 3 above)
Comment (QuanZhang-William) [reply pkg/apis/config/feature_flags.go:376]: Oh, sorry. I mean adding this tracking issue as comment in the codebase. As a developer, I could get confused by an added but never used struct (unless I missed the usage of it)

File: docs/developers/feature-versioning.md (hunk 1 above)
Comment (JeromeJu) [reply docs/developers/feature-versioning.md]: Thanks for the catch! updated.

File: pkg/apis/config/feature_flags.go (hunk 3 above)
Comment (JeromeJu) [reply pkg/apis/config/feature_flags.go:378]: Cool thanks. Done. Yeah that issue came into place just today.

**Review by QuanZhang-William — APPROVED** [review]

File: api_compatibility_policy.md [hunk 4]
```diff
@@ -90,7 +90,22 @@ For more information on support windows, see the [deprecations table](./docs/dep
...
+
+| Releases               | Features                      | API-driven Features Validation     |
+| ---------------------- | ----------------------------- | ---------------------------------- |
+| Prior to v0.53.0       | All features                  | `enable-api-fields` **only** |
+| After v0.53.0        | [Existing features](#snapshot-of-existent-beta-and-alpha-api-driven-features-as-of-v0530httpsgithubcomtektoncdpipelinereleasestagv0530) till v0.53.0 | `enable-api-fields`            |
+| After v0.53.0        | New features after v0.53.0      |  Per-feature flags                 |
+| ...                    |                               |                                    |
+| Future release       | All [existing features](#snapshot-of-existent-beta-and-alpha-api-driven-features-as-of-v0530httpsgithubcomtektoncdpipelinereleasestagv0530) stabilized or deprecated | Sunset ~~`enable-api-fields`~~ |
```
Comment (afrittoli) [inline api_compatibility_policy.md:103]: "All existing features" will sound confusing in future, despite the link. 
I find the table a bit difficult to parse, what about this?

| Releases               | Global flag `enable-api-fields`                       | Per-feature flags     |
| ---------------------- | ----------------------------- | ---------------------------------- |
| Prior to v0.53.0       | All API features                  |  Alpha/beta  non-API features only |
| After v0.53.0        | [Alpha/beta features in v0.53.0](#snapshot-of-existent-beta-and-alpha-api-driven-features-as-of-v0530httpsgithubcomtektoncdpipelinereleasestagv0530) | Alpha/beta  non-API features and API features introduced after v0.53.0  |
| All [alpha/beta features in v0.53.0](#snapshot-of-existent-beta-and-alpha-api-driven-features-as-of-v0530httpsgithubcomtektoncdpipelinereleasestagv0530) become stable or are removed |    Sunset ~~`enable-api-fields`~~   |  All alpha/beta API and NON-API features |

File: api_compatibility_policy.md [hunk 5]
```diff
@@ -90,7 +90,22 @@ For more information on support windows, see the [deprecations table](./docs/dep
...
+| After v0.53.0        | New features after v0.53.0      |  Per-feature flags                 |
+| ...                    |                               |                                    |
+| Future release       | All [existing features](#snapshot-of-existent-beta-and-alpha-api-driven-features-as-of-v0530httpsgithubcomtektoncdpipelinereleasestagv0530) stabilized or deprecated | Sunset ~~`enable-api-fields`~~ |  
+| ...                  | All features                    | Per-feature flags **only**         |
+
+With per-feature flags, cluster operators are able to enable a single new feature with their dedicated feature flags. For instructions on how to add a per-feature flag, please check the [developer feature versioning guide](./docs/developers/feature-versioning.md#per-feature-flag).
+
+Note that the `enable-api-fields` feature flag will continue to validate all existing [beta](https://github.com/tektoncd/pipeline/blob/release-v0.52.x/docs/additional-configs.md#beta-features) and [alpha](https://github.com/tektoncd/pipeline/blob/release-v0.52.x/docs/additional-configs.md#alpha-features) features till [v0.53.0](https://github.com/tektoncd/pipeline/tree/release-v0.52.x):
```
Comment (afrittoli) [inline api_compatibility_policy.md:108]: The links point to v0.52 - is that intentional? If so we should talk about "prior to v0.53".
The last link should definitely point to v0.53 though

Comment (afrittoli) [reply api_compatibility_policy.md:108]: I would rephrase a bit:

> Note that the `enable-api-fields` feature flag will continue to validate all features that were [beta](https://github.com/tektoncd/pipeline/blob/release-v0.52.x/docs/additional-configs.md#beta-features) and [alpha](https://github.com/tektoncd/pipeline/blob/release-v0.52.x/docs/additional-configs.md#alpha-features) prior to [v0.53.0](https://github.com/tektoncd/pipeline/tree/release-v0.53.x):

File: docs/developers/feature-versioning.md [hunk 6]
```diff
@@ -0,0 +1,166 @@
+# Feature Versioning
+
+The stability levels of features (feature versioning) are independent of CRD API versions (API versioning).
```
Comment (afrittoli) [inline docs/developers/feature-versioning.md:3]: what does "API versioning" in brackets mean here? Did you perhaps intend to link to the "api-versioning.md"?

File: docs/developers/feature-versioning.md [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+
+## Adding feature gates for API-driven features
+API-driven features are features that are enabled via a specific field in pipeline API. They comply to the [feature gates](../../api_compatibility_policy.md#feature-gates) and the [feature graduation process](../../api_compatibility_policy.md#feature-graduation-process) specified in the [API compatibility policy](../../api_compatibility_policy.md). For example, [remote tasks](https://github.com/tektoncd/pipeline/blob/454bfd340d102f16f4f2838cf4487198537e3cfa/docs/taskruns.md#remote-tasks) is an API-driven feature.
```
Comment (afrittoli) [inline docs/developers/feature-versioning.md:6]: NIT: I think "API-driven features" are features that are "accessed" or "consumed" via API, rather than "enabled".
Enabled is already associated with the flag, and I would find it consuming to use the same term.

File: docs/developers/feature-versioning.md [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+
+4. Add `example-new-feature: false` to [the pipeline configMap](./../../config/config-feature-flags.yaml) with a release note.
+
+5. Update documentations for the new alpha feature at [alpha-stability-level](./../additional-configs.md#alpha-features).
+
+### `enable-api-fields`
+
+We've introduced a feature-flag called `enable-api-fields` to the
+[config-feature-flags.yaml file](../../config/config-feature-flags.yaml)
+deployed as part of our releases.
```
Comment (afrittoli) [inline docs/developers/feature-versioning.md:44]: I find this confusing, it sounds like something newly introduced. 
I agree we need to keep the `enable-api-fields` docs, but we should mention here too that this flag is deprecated and shall not be used for new features.

**Review by afrittoli — COMMENTED** [review]
Thanks @JeromeJu - I left some comments.

File: api_compatibility_policy.md (hunk 5 above)
Comment (JeromeJu) [reply api_compatibility_policy.md]: Thanks! Updated!

File: docs/developers/feature-versioning.md (hunk 7 above)
Comment (JeromeJu) [reply docs/developers/feature-versioning.md:6]: Done!

File: api_compatibility_policy.md [hunk 9]
```diff
@@ -90,7 +90,22 @@ For more information on support windows, see the [deprecations table](./docs/dep
...
+Stability levels of feature gates are independent from CRD apiVersions. 
+
+[TEP0138](https://github.com/tektoncd/community/blob/main/teps/0138-decouple-api-and-feature-versioning.md) has introduced per-feature flags for new API-driven features and the migration plan for `enable-api-fields`. Please refer to the table below for the API-driven features validation changes:
+
+| Releases               | Features                      | API-driven Features Validation     |
+| ---------------------- | ----------------------------- | ---------------------------------- |
+| Prior to v0.53.0       | All features                  | `enable-api-fields` **only** |
+| After v0.53.0        | [Existing features](#snapshot-of-existent-beta-and-alpha-api-driven-features-as-of-v0530httpsgithubcomtektoncdpipelinereleasestagv0530) till v0.53.0 | `enable-api-fields`            |
+| After v0.53.0        | New features after v0.53.0      |  Per-feature flags                 |
+| ...                    |                               |                                    |
```
[the remark this answers is not in the document]
Comment (JeromeJu) [reply api_compatibility_policy.md]: TY @afrittoli updated!

File: api_compatibility_policy.md (hunk 4 above)
Comment (JeromeJu) [reply api_compatibility_policy.md]: That makes sense! I made some minor changes to it because behavioural flags will retain their behaviour. (I also added a note below)

File: docs/developers/feature-versioning.md (hunk 6 above)
Comment (JeromeJu) [reply docs/developers/feature-versioning.md]: Ah yes I've linked it to the doc.

File: docs/developers/feature-versioning.md (hunk 8 above)
Comment (JeromeJu) [reply docs/developers/feature-versioning.md]: Updated:) these were previous doc, I've edited them

Comment (JeromeJu) [reply docs/developers/feature-versioning.md]: Thanks for the catch!

File: api_compatibility_policy.md (hunk 4 above)
Comment (JeromeJu) [reply api_compatibility_policy.md]: Also updated the links at https://github.com/tektoncd/community/pull/1091.

**Review by afrittoli — COMMENTED** [review]
/lgtm
